appmap-java-maven-plugin/src/main/java/com/appland/appmap/LoadJavaAppMapAgentMojo.java

```diff
@@ -0,0 +1,68 @@
+package com.appland.appmap;
+
+/*
+ * Copyright 2001-2005 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.maven.plugin.AbstractMojo;
+import org.apache.maven.plugin.MojoExecutionException;
+import com.appland.appmap.record.IRecordingSession;
+import com.appland.appmap.record.Recorder;
+
+import java.io.File;
+
+/**
+ * Goal which register a appmap java recording agent, right before the test execution begins.
+ *
+ * @goal prepare-agent
+ * @phase test-compile
+ */
+public class LoadJavaAppMapAgentMojo extends AbstractMojo {
+    /**
+     * Directory where reports will be generated.
+     *
+     * @parameter expression="${project.build.directory}"
+     *
+     */
+    private File outputDirectory;
+
+    /**
+     * Configuration file for the java agent @default /appmap.yml.
+     *
+     * @parameter expression="${project.build.directory}"
+     *
+     */
+    private File configFile;
+
+    /**
+     * Java Agent Debug Option.
+     *
+     * @parameter expression="${project.build.directory}
+     *
+     */
+    private Boolean debug = false;
+
+    public void execute()
+            throws MojoExecutionException {
+        getLog().info("Initializing AppLand AppMap Java Recorder");
+        try {
+            final IRecordingSession.Metadata metadata = new IRecordingSession.Metadata();
+            Recorder.getInstance().start(metadata);
+        }catch (Exception e) {
+            getLog().error("Error initializing AppLand AppMap Java Recorder");
+            e.printStackTrace();
+        }
```

Comment:
@marcuss this is still not correct. Take a look at executeMojo() in [https://github.com/jacoco/jacoco/blob/master/jacoco-maven-plugin/src/org/jacoco/maven/AbstractAgentMojo.java](https://github.com/jacoco/jacoco/blob/master/jacoco-maven-plugin/src/org/jacoco/maven/AbstractAgentMojo.java). 

You should not start the recorder in the plugin. Instead, you need to load a resource from the appmap.jar; which in turn will activate the instrumentation (see the manifest of appmap.jar and the Premain-class attribute). jacoco's executeMojo() does it by calling this method:
	AgentOptions createAgentOptions() {
		final AgentOptions agentOptions = new AgentOptions();

AgentOptions is a class in jacoco's agent jar file. When this class is instantiated, class loading of the jar file initializes the jacoco instrumentation. Since you're loading and executing bytecode from the appmap.jar, the instrumentation gets initialized too, but you should not be starting the recording from the plugin.

Reply:
Corrected to work as expected.

Please be aware this is just a draft PR, and this first version needs A LOT of refactoring, just pushing changes so I can show the day-to-day progress.